cores/arduino/stm32/dma.c

```diff
@@ -0,0 +1,293 @@
+/**
+  ******************************************************************************
+  * @file    dma.c
+  * @author  MCD Application Team, Wi6labs
+  * @brief   dma service.
+  ******************************************************************************
+
+Copyright (c) <2017>, <Wi6labs>
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of the wi6labs nor the
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL WI6LABS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  ******************************************************************************
+  */
+
+#include "stm32_def.h"
+#include "dma.h"
+#include "analog.h"
+
+#ifdef __cplusplus
+ extern "C" {
+#endif
+
+extern ADC_HandleTypeDef AdcHandleTab[ADC_NUM];
+
+static bool dmaTab[DMA_NUMBER][CHANNEL_NUMBER];
+
+/**
+  * @brief  This function will init the dma and channel corresponding 
+  *         to the wanted peripheral
+  * @param  periph_handle : pointer to the peripheral HandleTyeDef
+  * @param  periph        : the peripheral type to use
+  * @retval none
+  */
+void dma_allocate(void *periph_handle, Peripheral periph, void **channel, IRQn_Type *irqHandler, uint8_t *indice)
+{
```

Comment:
Why passing a periph handle and a peripheral type ? 
Can't we pass pass the peripheral Instance directly ? 
dma_allocate(uint32_t *Instance, void **channel, IRQn_Type *irqHandler, uint8_t *indice);
as we know that Instance is actually the peripheral memory address.

Reply:
If you don't have the peripheral type, how did you want to cast your void pointer ? 
hadc = (ADC_HandleTypeDef*)periph_handle; Here I need to know i'm working on a ADC

Reply:
Why do you want to cast it?
You just need to compare 2 void* Instances 
Use void* as entries in the DmaMap_DMA and use the same Instance address as void* type for comparing.

Reply:
By doing this, you make the assumption that instance is always at the beginning of the peripheral struct. Is it always true for all peripheral? Can we be sure it will not change in the future?

Reply:
No. Here I'm proposing to change the function to:
dma_allocate(void *Instance, void **channel, IRQn_Type *irqHandler, uint8_t *indice)
So the caller will pass the Instance as a parameter. 
As far as I see, you are only using hadc->Instance in this function, so you don't need the hadc handle, you just need the instance.

Reply:
OK my bad, i didn't see you change the prototype. Yes ok, with that it can work, only pass the instance and not the entire struct.